File: desktopApp/core/src/main/java/si/um/feri/maprri/raster/backendCalls/FetchGardens.java
package si.um.feri.maprri.raster.backendCalls;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Net;
import com.badlogic.gdx.net.HttpRequestBuilder;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.Json;
import si.um.feri.maprri.raster.Garden;

public class FetchGardens {

    public interface GardensCallback {
        void onSuccess(Array<Garden> gardens);
        void onError(Throwable t);
    }


    public static void getAllGardens(String backendUrl, GardensCallback callback) {
        HttpRequestBuilder requestBuilder = new HttpRequestBuilder();
        Net.HttpRequest request = requestBuilder.newRequest()
                .method(Net.HttpMethods.GET)
                .url(backendUrl + "/garden/")
                .build();

        Gdx.net.sendHttpRequest(request, new Net.HttpResponseListener() {
            @Override
            public void handleHttpResponse(Net.HttpResponse httpResponse) {
                try {
                    String json = httpResponse.getResultAsString();
//                    System.out.println("Prejeli JSON: " + json);

                    Json gdxJson = new Json();

                    if (json.trim().startsWith("[")) {
                        Array<Garden> gardensArray = gdxJson.fromJson(Array.class, Garden.class, json);
                        callback.onSuccess(gardensArray);
                    } else {
                        System.err.println("Napaka od backend-a: " + json);
                        callback.onError(new Exception("Invalid JSON from backend: " + json));
                    }

                } catch (Exception e) {
                    e.printStackTrace();
                    callback.onError(e);
                }
            }

            @Override
            public void failed(Throwable t) {
                t.printStackTrace();
                callback.onError(t);
            }

            @Override
            public void cancelled() {
                callback.onError(new Exception("Request cancelled"));
            }
        });

    }

}

File: mobileApp/app/src/main/java/si/um/feri/mobilegarden/utils/FetchGardens.kt
package si.um.feri.mobilegarden.utils

import android.util.Log
import com.google.gson.Gson
import si.um.feri.mobilegarden.models.Garden
import java.net.URL

object FetchGardens {

    interface GardensCallback {
        fun onSuccess(gardens: List<Garden>)
        fun onError(e: Exception)
    }

    fun getAllGardens(backendUrl: String, callback: GardensCallback) {
        Thread {
            try {
                val url = "$backendUrl/garden/"
                Log.d("FetchGardens", "Calling URL: $url")

                val response = URL(url).readText()
                Log.d("FetchGardens", "Response: $response")

                val gson = Gson()
                val listType = object : com.google.gson.reflect.TypeToken<List<Garden>>() {}.type
                val gardens: List<Garden> = gson.fromJson(response, listType)

                callback.onSuccess(gardens)

            } catch (e: Exception) {
                Log.e("FetchGardens", "ERROR fetching gardens", e)
                callback.onError(e)
            }
        }.start()
    }

}
